examples/fabric-admin/rpc/RpcClient.cpp

```diff
@@ -0,0 +1,168 @@
+
+// Function to process incoming packets
+void ProcessPackets()
+{
+    std::array<std::byte, kMaxTransmissionUnit> inputBuf;
+    pw::hdlc::Decoder decoder(inputBuf);
+
+    while (true)
+    {
+        std::array<std::byte, kMaxTransmissionUnit> data;
+        auto ret = rpcSocketStream.Read(data);
+        if (!ret.ok())
+        {
+            if (ret.status() == pw::Status::OutOfRange())
+            {
+                // Handle remote disconnect
+                rpcSocketStream.Close();
+                return;
+            }
+            continue;
+        }
+
+        for (std::byte byte : ret.value())
+        {
+            auto result = decoder.Process(byte);
+            if (!result.ok())
+            {
+                // Wait for more bytes that form a complete packet
+                continue;
+            }
+            pw::hdlc::Frame & frame = result.value();
+            if (frame.address() != pw::hdlc::kDefaultRpcAddress)
+            {
+                // Wrong address; ignore the packet
+                continue;
+            }
+
+            client.ProcessPacket(frame.data()).IgnoreError();
+        }
+    }
+}
+
+template <typename CallType>
+CHIP_ERROR WaitForResponse(CallType & call)
+{
+    if (!call.active())
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    // Wait for the response or timeout
+    uint32_t elapsedTimeMs     = 0;
+    const uint32_t sleepTimeMs = 100;
+
+    while (call.active() && elapsedTimeMs < kRpcTimeoutMs)
+    {
+        usleep(sleepTimeMs * 1000);
+        elapsedTimeMs += sleepTimeMs;
+    }
+
+    if (elapsedTimeMs >= kRpcTimeoutMs)
+    {
+        fprintf(stderr, "RPC Response timed out!");
+        return CHIP_ERROR_TIMEOUT;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void AddDeviceResponse(const pw_protobuf_Empty & response, pw::Status status)
+{
+    if (status.ok())
+    {
+        printf("RPC call succeeded\n");
+    }
+    else
+    {
+        printf("RPC call failed with status: %d\n", status.code());
+    }
+}
+
+} // namespace
+
+CHIP_ERROR InitRpcClient(uint16_t rpcServerPort)
+{
+    if (rpcSocketStream.Connect(rpcServerAddress, rpcServerPort) != PW_STATUS_OK)
+    {
+        ChipLogError(NotSpecified, "Failed to connect the Fabric-Admin");
+        return CHIP_ERROR_NOT_CONNECTED;
+    }
+
+    ChipLogProgress(NotSpecified, "Connectted to the Fabric-Admin\n");
+
+    // Start a thread to process incoming packets
+    std::thread packet_processor(ProcessPackets);
+    packet_processor.detach();
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AddSynchronizedDevice(chip::NodeId nodeId)
+{
+    ChipLogProgress(NotSpecified, "AddSynchronizedDevice");
+
+    FabricBridgeClient fabric_bridge_client(client, kDefaultChannelId);
+    chip_rpc_SynchronizedDevice device;
+    device.node_id = nodeId;
+
+    // The RPC will remain active as long as `call` is alive.
+    auto call = fabric_bridge_client.AddSynchronizedDevice(device, AddDeviceResponse);
+    return WaitForResponse(call);
```

Comment:
Is this blocking the main MatterEventLoop here? If so, is that what we want to do

Reply:
Changed to asynchronous mode